pageserver/src/storage_sync.rs

```diff
@@ -129,1743 +150,722 @@
-            }
-            DownloadMarker::Nothing
-        }
-        .instrument(info_span!("download_timeline_data")),
-    );
-
-    if let Some(mut delete_data) = batch.delete {
-        if upload_result.is_some() {
-            match validate_task_retries(delete_data, max_sync_errors)
-                .instrument(info_span!("retries_validation"))
-                .await
-            {
-                ControlFlow::Continue(new_delete_data) => {
-                    delete_timeline_data(
-                        conf,
-                        (storage.as_ref(), &index, sync_queue),
-                        sync_id,
-                        new_delete_data,
-                        sync_start,
-                        "delete",
-                    )
-                    .instrument(info_span!("delete_timeline_data"))
-                    .await;
+                UploadOp::Delete(_) => {
+                    upload_queue.num_inprogress_deletions += 1;
+                }
+                UploadOp::Barrier(sender) => {
+                    sender.send_replace(());
+                    continue;
                 }
-                ControlFlow::Break(failed_delete_data) => {
-                    if let Err(e) = update_remote_data(
-                        conf,
-                        storage.as_ref(),
-                        &index,
-                        sync_id,
-                        RemoteDataUpdate::Delete(&failed_delete_data.data.deleted_layers),
+            };
+
+            // Assign unique ID to this task
+            upload_queue.task_counter += 1;
+            let task_id = upload_queue.task_counter;
+
+            // Add it to the in-progress map
+            let task = Arc::new(UploadTask {
+                task_id,
+                op: next_op,
+                retries: AtomicU32::new(0),
+            });
+            upload_queue
+                .inprogress_tasks
+                .insert(task.task_id, Arc::clone(&task));
+
+            // Spawn task to perform the task
+            let self_rc = Arc::clone(self);
+            let _handle = self.runtime.spawn(async move {
+                self_rc.perform_upload_task(task).await;
+            });
+
+            // Loop back to process next task
+        }
+    }
+
+    ///
+    /// Perform an upload task.
+    ///
+    /// The task is in the `inprogress_tasks` list. This function will try to
+    /// execute it, retrying forever. On successful completion, the task is
+    /// removed it from the `inprogress_tasks` list, and any next task(s) in the
+    /// queue that were waiting by the completion are launched.
+    ///
+    async fn perform_upload_task(self: &Arc<Self>, task: Arc<UploadTask>) {
+        // Loop to retry until it completes.
+        loop {
+            let upload_result: anyhow::Result<()> = match task.op {
+                UploadOp::UploadLayer(ref path) => {
+                    upload::upload_timeline_layer(&self.storage_impl, path).await
```

Comment:
We have to upload remote index part after that, to add a new file path into that index, are we doing that somewhere?
If not, we'll loose any new files, since we're not calling S3 list anywhere.